hub/core/transform/test_transform.py

```diff
@@ -0,0 +1,97 @@
+from hub.api.dataset import Dataset
+from hub.core.transform.transform import transform  # type: ignore
+import numpy as np
+from hub.core.tests.common import parametrize_all_dataset_storages
+
+
+def fn1(i, mul=1, copy=1):
+    d = {}
+    d["image"] = np.ones((337, 200)) * i * mul
+    d["label"] = np.ones((1,)) * i * mul
+    return [d for _ in range(copy)]
+
+
+def fn2(sample, mul=1, copy=1):
+    d = {}
+    d["image"] = sample["image"] * mul
+    d["label"] = sample["label"] * mul
+    return [d for _ in range(copy)]
+
+
+def fn3(i, mul=1, copy=1):
+    d = {}
+    d["image"] = np.ones((1310, 2087)) * i * mul
+    d["label"] = np.ones((13,)) * i * mul
+    return [d for _ in range(copy)]
+
+
+@parametrize_all_dataset_storages
+def test_transform_hub_dataset(ds):
+    with Dataset("./test/transform_hub_in") as data_in:
+        data_in.create_tensor("image")
+        data_in.create_tensor("label")
+        for i in range(100):
+            data_in.image.append(i * np.ones((100, 100)))
+            data_in.label.append(i * np.ones((1,)))
+    data_in = Dataset("./test/transform_hub_in")
+    ds_out = ds
+    ds_out.create_tensor("image")
```

Comment:
I don't like that we have to do this. What's wrong with inferring the tensors from samples?

Could we:
- Transform the first sample
- Deduce the necessary output tensors
- Create the necessary tensors on `ds_out` if they don't exist
?

Reply:
The issue with this is that separate processes are not communicating with each other till the very end. This means that in the end, we could end up with datasets with different 'schemas' which are impossible to merge as transforms are generated sample-wise. This would cause the error at the very end, after transforms are complete, instead of when an incorrect sample is detected.
Doing it the way we have it right now, all samples are compared to tensors as they are generated and any mismatch is thrown immediately.
I don't think it's a huge issue because it essentially makes users think about transforms as something that returns some exact tensors for each sample, instead of arbitrary dicts. It is good to have some restrictions imo. I can remove these restrictions but they will lead to non-sensical datasets being generated using transforms that won't be very useful.

Moreover, doing it this way gives us more power as well. Right now transforms also support append functionality which means that we can periodically return a transform job to add new samples. The only requirement for that is that the dict outputs of the transform match exactly with ds_out's tensors and all the tensors of ds_out  have equal length at the start.

Reply:
@AbhinavTuli Understood. I still think that in the simple case we should have some sugar to sweeten up this API for the users, but this is fine for now.